Application Connectivity capability description (#131)
kyma-project/community: capabilities/application-connectivity.md

---
displayName: Application Connectivity
epicsLabels:
  - area/application-connector
---

## Scope

> The Application Connectivity capability supports the extensibility and integration between applications in a cloud-native way using the best available solutions.

Reviewer: For me, the scope is not clear. I would expect something that explains what problem Application Connectivity solves and how.

    * Registered entities are integrated with Service Catalog
     
* Orchestration

    * The orchestration functionality allows automatic binding between Kyma runtime and application registered by the customer.
    * The orchestration is tightly coupled with the API registry and discovery.
    
* Events integration
    * The event integration functionality provides required middleware for delivery of the business events to Kyma.
    * The support for the delivery guarantee, monitoring, and tracing is added.
    
* Access to the registered APIs
    * The access to API exposed by an integrated application is provided, and all requests are proxied by the delivered proxy service.
    * The proxy service is handling the authentication and authorization, integration with monitoring and tracing. 
    * The Service Catalog binding controls the access to the proxy service and therefore to the API. The development effort is reduced to the required minimum, and all the boiler-plate code is packed into connectors.
    

> * Connectors

Author: The description has been improved.

>     * The set of connectors ensures that plumbing between an external application and SAP CP XF is possible in all cases where the use of standard functionality is not feasible. 

Reviewer: What is standard functionality? Connectors are not part of standard functionality? 
Please remove reference to SAP CP XF.

---
displayName: Application Connectivity
epicsLabels:
  - area/application-connector
---

## Scope

The Application Connectivity capability supports the extensibility and integration between applications in a cloud-native way using the best available solutions.
It provides the integration patterns and necessary tooling for establishing the trusted connection between Kyma and third-party applications.

The goal is to enable the integration of various systems in a coherent way, where the management and usage of APIs and Events are standardized and allows natural extension of such systems in the cloud-native fashion.
The unification of the connected system enables the fast development which unlocks the brand new possibilities of extending and customizing the existing solution in the modern, cloud-based, way.

> The new way of scalability of the system is possible. The event mechanism and simple access to exposed API is a foundation for moving the workload from the legacy system to the Kubernetes.

Reviewer 2: ```suggestion
A new way of system scalability is available. The event mechanism and simple access to the exposed API is a foundation for moving the workload from a legacy system to Kubernetes.
```

> An additional benefit is that you can mesh different systems using the language of your choice. 

Reviewer 2: ```suggestion
An additional benefit is that you can connect different systems using the programming language of your choice. 
```

The integration must be as simple as possible.

## Vision

> * API Registry and Discoverability

Reviewer: It is more Application Registry not only API Registry. App = APIs + Events

    * The registration contains the configuration of the endpoints together with required security credentials, documentation of the API and Event catalog based on open standards like OpenAPI and AsyncAPI, and additional documentation.
    * Registered entities are integrated with Service Catalog
     
* Orchestration

>     * The orchestration functionality allows automatic binding between Kyma runtime and application registered by the customer.

Reviewer 2: ```suggestion
    * The orchestration functionality allows for an automatic binding between the Kyma runtime and the application registered by the customer.
```

> * Orchestration

Reviewer: Orchestration is a little bit blurry. I would like to see: it is possible to write the cloud native, scallable extension that can consume (or be triggered by) any event coming from the connected application and can call any connected application API without additional configuration effort. The App registry information is available for developers in the tooling.

Author: I have delivered a different description. Still a little bit blurry. It requires further discussion.

>     * The orchestration is tightly coupled with the API registry and discovery.

Reviewer 2: ```suggestion
    * The orchestration is tightly coupled with the API registry and service discovery.
```

    
* Events integration
    * The event integration functionality provides required middleware for delivery of the business events to Kyma.

>     * The support for the delivery guarantee, monitoring, and tracing is added.

Reviewer 2: ```suggestion
    * The support for delivery guarantee, monitoring, and tracing is added.
```

>     *It's built the trusted relationship between all connected applications.

Reviewer: ```suggestion
    * It's built the trusted relationship between all connected applications.
```

    
* Access to the registered APIs

>     * The access to API exposed by an integrated application is provided, and all requests are proxied by the delivered proxy service.

Reviewer 2: ```suggestion
    * The access to the APIs exposed by an integrated application is provided, and all requests are proxied by the delivered proxy service.
```

>     * The proxy service is handling the authentication and authorization, integration with monitoring and tracing. 

Reviewer 2: ```suggestion
    * The proxy service is handling the authentication and authorization, integration with monitoring, and tracing. 
```

    * The Service Catalog binding controls the access to the proxy service and therefore to the API. The development effort is reduced to the required minimum, and all the boiler-plate code is packed into connectors.
    
* Connectors

>     * In the case of integration with the legacy system, which is not exposed using open standards, like REST, or if it is hidden behind a network firewall, the connectors will be provided.

Reviewer 2: ```suggestion
    * In the case of integration with the legacy system, which is not exposed using open standards such as REST, or if it is hidden behind a network firewall, the connectors will be provided.
```

>     * The connector ensures that a legacy system will be exposed in the same way as other systems, and the required translation of the API calls or events will be done. 

Reviewer 2: ```suggestion
    * The connector ensures that a legacy system will be exposed in the same way as other systems, and the required translation of the API calls or events will be handled. 
```

    * The palette of provided connectors should be kept to the required minimum and wherever possible the industry standards must be used.

* Security

>     * The security of the integration is one of the top priority. The trusted relationship between connected application must be ensured, including the authentication and authorization.

Reviewer 2: ```suggestion
    * The security of the integration is one of the top priorities. The trusted relationship between connected application must be ensured, including authentication and authorization.
```

>     * Various standard security mechanisms are provided to ensure the identity of the application, access to the secure resources, in both communication directions, no matter if it is asynchronous communication using events or synchronous communication using REST calls. The support for OAuth, Basic auth, client certificates, CSRF tokes and more must be provided.

Reviewer 2: ```suggestion
    * Various standard security mechanisms are provided to ensure the identity of the application, access to the secure resources, in both communication directions. No matter if it is asynchronous communication using events, or synchronous communication using REST calls. The support for OAuth, Basic Auth, client certificates, CSRF tokens, etc. more must be provided.
```
